File: tools/mcp_converter.py
#!/usr/bin/env python3
"""
Convertidor de MCP (Model Context Protocol) a ATDF (Agent Tool Description Format).

Este módulo proporciona funciones para convertir herramientas desde el formato MCP
utilizado por algunos modelos de lenguaje, al formato ATDF estándar.
"""

import json
import logging
from pathlib import Path
from typing import Any, Dict, List, Optional, Union

from .converter import convert_to_enhanced, save_tool

# Referencia a funciones existentes del proyecto
from .validator import validate_tool_smart

# Configuración de logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger("mcp_converter")


def mcp_to_atdf(
    mcp_tool: Dict[str, Any], enhanced: bool = False, author: str = "MCP Converter"
) -> Dict[str, Any]:
    """
    Convierte una herramienta MCP a formato ATDF.

    Args:
        mcp_tool (Dict): Objeto JSON de una herramienta MCP.
        enhanced (bool): Si es True, convierte al formato mejorado de ATDF.
        author (str): Autor para los metadatos en formato mejorado.

    Returns:
        Dict: Objeto ATDF válido.

    Raises:
        ValueError: Si faltan campos obligatorios o la validación falla.
    """
    # Validar campos obligatorios en MCP
    required_mcp_fields = ["name", "inputSchema"]
    for field in required_mcp_fields:
        if field not in mcp_tool:
            raise ValueError(f"Campo obligatorio '{field}' no encontrado en MCP")

    # Mapear campos MCP → ATDF
    atdf_tool = {
        "schema_version": "1.0.0",
        "tool_id": mcp_tool["name"],
        "description": mcp_tool.get("description", "Sin descripción proporcionada"),
        "when_to_use": _extract_when_to_use(mcp_tool),
        "how_to_use": {
            "inputs": _parse_inputs(mcp_tool["inputSchema"]),
            "outputs": _default_outputs(),
        },
    }

    # Validar con el esquema ATDF
    if not validate_tool_smart(atdf_tool):
        raise ValueError("[✗] Validación ATDF fallida")

    logger.info(f"[✓] {atdf_tool['tool_id']} validado exitosamente como ATDF")

    # Convertir a formato mejorado si se solicita
    if enhanced:
        atdf_tool = convert_to_enhanced(atdf_tool, author=author, extract_language=True)
        logger.info(f"Herramienta convertida a formato ATDF mejorado")

    return atdf_tool


def _extract_when_to_use(mcp_tool: Dict[str, Any]) -> str:
    """
    Extrae o genera contexto de uso desde MCP.

    Args:
        mcp_tool: Objeto JSON de la herramienta MCP.

    Returns:
        String con el contexto de uso para ATDF.
    """
    # Priorizar anotaciones MCP si existen
    if "annotations" in mcp_tool:
        context_hints = [
            f"Uso: {hint}"
            for key, hint in mcp_tool["annotations"].items()
            if key in ["context", "usage", "purpose"]
        ]
        if context_hints:
            return " ".join(context_hints)

    # Generar mensaje por defecto
    return (
        f"Usar cuando se requiera {mcp_tool.get('description', '').lower().strip('.')}"[
            :200
        ]
    )


def _parse_inputs(input_schema: Dict[str, Any]) -> List[Dict[str, Any]]:
    """
    Convierte JSON Schema de MCP a formato inputs de ATDF.

    Args:
        input_schema: Esquema de entrada MCP.

    Returns:
        Lista de parámetros en formato ATDF.
    """
    inputs = []

    # Procesar las propiedades del esquema
    for param_name, param_def in input_schema.get("properties", {}).items():
        input_param = {
            "name": param_name,
            "type": param_def.get("type", "unknown"),
            "description": param_def.get("description", f"Parámetro {param_name}"),
        }

        # Marcar como requerido si está en la lista required
        if "required" in input_schema and param_name in input_schema["required"]:
            input_param["required"] = True

        inputs.append(input_param)

    return inputs


def _default_outputs(
    success_msg: str = None, failure_codes: List[Dict[str, str]] = None
) -> Dict[str, Any]:
    """
    Genera estructura de salidas para ATDF.

    Args:
        success_msg: Mensaje personalizado de éxito.
        failure_codes: Lista de códigos y descripciones de error.

    Returns:
        Estructura de outputs válida para ATDF.
    """
    # Mensaje de éxito por defecto o personalizado
    if not success_msg:
        success_msg = "Operación completada exitosamente"

    # Códigos de error por defecto o personalizados
    if not failure_codes:
        failure_codes = [
            {"code": "invalid_input", "description": "Entrada inválida o incompleta"},
            {"code": "tool_error", "description": "Error interno de la herramienta"},
        ]

    return {"success": success_msg, "failure": failure_codes}


def convert_mcp_file(
    input_file: Union[str, Path],
    output_file: Optional[Union[str, Path]] = None,
    format: str = "json",
    enhanced: bool = False,
) -> Optional[Dict[str, Any]]:
    """
    Convierte un archivo de herramienta MCP a formato ATDF.

    Args:
        input_file: Ruta al archivo MCP a convertir.
        output_file: Ruta donde guardar el resultado. Si es None, solo retorna el resultado.
        format: Formato de salida ("json" o "yaml").
        enhanced: Si es True, convierte al formato mejorado.

    Returns:
        Dict con la herramienta convertida o None si hay errores.
    """
    try:
        input_path = Path(input_file)

        # Verificar existencia del archivo
        if not input_path.exists():
            logger.error(f"Archivo no encontrado: {input_path}")
            return None

        # Cargar archivo MCP
        with open(input_path, "r", encoding="utf-8") as f:
            if input_path.suffix.lower() == ".json":
                mcp_data = json.load(f)
            else:
                logger.error(f"Formato no soportado: {input_path.suffix}")
                return None

        # Convertir a ATDF
        atdf_result = mcp_to_atdf(mcp_data, enhanced=enhanced)

        # Guardar si se especificó archivo de salida
        if output_file:
            save_tool(atdf_result, output_file, format=format)
            logger.info(f"Herramienta guardada en {output_file}")

        return atdf_result

    except Exception as e:
        logger.error(f"Error en la conversión: {str(e)}")
        return None


# Ejemplo de uso independiente
if __name__ == "__main__":
    # Ejemplo MCP: Herramienta 'fetch'
    mcp_tool_example = {
        "name": "fetch",
        "description": "Recupera contenido de una URL",
        "annotations": {"context": "Cuando necesites obtener datos de una página web"},
        "inputSchema": {
            "type": "object",
            "properties": {
                "url": {"type": "string", "description": "URL a recuperar"},
                "raw": {
                    "type": "boolean",
                    "description": "Devolver HTML crudo (opcional)",
                },
            },
            "required": ["url"],
        },
    }

    try:
        # Convertir a ATDF
        atdf_result = mcp_to_atdf(mcp_tool_example)

        # Imprimir resultado
        print("\n✅ ATDF generado:")
        print(json.dumps(atdf_result, indent=2))

        # Opcionalmente convertir y guardar en formato mejorado
        enhanced_result = mcp_to_atdf(mcp_tool_example, enhanced=True)
        save_tool(enhanced_result, "./examples/fetch_enhanced.json")

    except Exception as e:
        print(f"[ERROR] {str(e)}")

File: js/src/MCPConverter.js
/**
 * Convertidor de MCP (Model Context Protocol) a ATDF (Agent Tool Description Format).
 *
 * Este módulo proporciona funciones para convertir herramientas desde el formato MCP
 * utilizado por algunos modelos de lenguaje, al formato ATDF estándar.
 */

const fs = require('fs');
const path = require('path');

/**
 * Convierte una herramienta MCP a formato ATDF.
 *
 * @param {Object} mcpTool - Objeto JSON de una herramienta MCP.
 * @param {Object} options - Opciones de conversión.
 * @param {boolean} options.enhanced - Si es true, añade campos adicionales del formato mejorado.
 * @param {string} options.author - Autor para los metadatos en formato mejorado.
 * @returns {Object} - Objeto ATDF válido.
 * @throws {Error} - Si faltan campos obligatorios.
 */
function mcpToAtdf(mcpTool, options = {}) {
  const { enhanced = false, author = 'MCP Converter' } = options;

  // Validar campos obligatorios en MCP
  const requiredMcpFields = ['name', 'inputSchema'];
  for (const field of requiredMcpFields) {
    if (!(field in mcpTool)) {
      throw new Error(`Campo obligatorio '${field}' no encontrado en MCP`);
    }
  }

  // Mapear campos MCP → ATDF
  const atdfTool = {
    schema_version: '1.0.0',
    tool_id: mcpTool.name,
    description: mcpTool.description || 'Sin descripción proporcionada',
    when_to_use: extractWhenToUse(mcpTool),
    how_to_use: {
      inputs: parseInputs(mcpTool.inputSchema),
      outputs: defaultOutputs(),
    },
  };

  // Añadir campos del formato mejorado si se solicita
  if (enhanced) {
    atdfTool.schema_version = '2.0.0';

    // Añadir sección de metadatos
    atdfTool.metadata = {
      version: '1.0.0',
      author: author,
      tags: ['converted', 'mcp'],
      category: 'auto_converted',
      created_at: new Date().toISOString().split('T')[0],
      updated_at: new Date().toISOString().split('T')[0],
    };

    // Añadir sección de prerequisitos
    atdfTool.prerequisites = {
      tools: [],
      conditions: [],
      permissions: [],
    };

    // Añadir sección de feedback
    atdfTool.feedback = {
      progress_indicators: ['Operation in progress'],
      completion_signals: ['Operation completed'],
    };

    // Añadir sección de ejemplos
    atdfTool.examples = [
      {
        goal: `Example use of ${mcpTool.name}`,
        input_values: {},
        expected_result: 'Success',
      },
    ];
  }

  console.log(`[✓] ${atdfTool.tool_id} validado exitosamente como ATDF`);
  return atdfTool;
}

/**
 * Extrae o genera contexto de uso desde MCP.
 *
 * @param {Object} mcpTool - Objeto JSON de la herramienta MCP.
 * @returns {string} - String con el contexto de uso para ATDF.
 */
function extractWhenToUse(mcpTool) {
  // Priorizar anotaciones MCP si existen
  if (mcpTool.annotations) {
    const contextKeys = ['context', 'usage', 'purpose'];
    const contextHints = [];

    for (const key of contextKeys) {
      if (mcpTool.annotations[key]) {
        contextHints.push(`Uso: ${mcpTool.annotations[key]}`);
      }
    }

    if (contextHints.length > 0) {
      return contextHints.join(' ');
    }
  }

  // Generar mensaje por defecto
  const description = mcpTool.description || '';
  return `Usar cuando se requiera ${description.toLowerCase().replace(/\.$/, '')}`.substring(
    0,
    200
  );
}

/**
 * Convierte JSON Schema de MCP a formato inputs de ATDF.
 *
 * @param {Object} inputSchema - Esquema de entrada MCP.
 * @returns {Array} - Lista de parámetros en formato ATDF.
 */
function parseInputs(inputSchema) {
  const inputs = [];

  // Procesar las propiedades del esquema
  const properties = inputSchema.properties || {};
  for (const [paramName, paramDef] of Object.entries(properties)) {
    const inputParam = {
      name: paramName,
      type: paramDef.type || 'unknown',
      description: paramDef.description || `Parámetro ${paramName}`,
    };

    // Marcar como requerido si está en la lista required
    if (inputSchema.required && inputSchema.required.includes(paramName)) {
      inputParam.required = true;
    }

    inputs.push(inputParam);
  }

  return inputs;
}

/**
 * Genera estructura de salidas para ATDF.
 *
 * @param {string} successMsg - Mensaje personalizado de éxito.
 * @param {Array} failureCodes - Lista de códigos y descripciones de error.
 * @returns {Object} - Estructura de outputs válida para ATDF.
 */
function defaultOutputs(successMsg = null, failureCodes = null) {
  // Mensaje de éxito por defecto o personalizado
  if (!successMsg) {
    successMsg = 'Operación completada exitosamente';
  }

  // Códigos de error por defecto o personalizados
  if (!failureCodes) {
    failureCodes = [
      { code: 'invalid_input', description: 'Entrada inválida o incompleta' },
      { code: 'tool_error', description: 'Error interno de la herramienta' },
    ];
  }

  return {
    success: successMsg,
    failure: failureCodes,
  };
}

/**
 * Convierte un archivo MCP a formato ATDF.
 *
 * @param {string} inputFile - Ruta al archivo MCP a convertir.
 * @param {string} outputFile - Ruta donde guardar el resultado. Si es null, solo retorna el resultado.
 * @param {Object} options - Opciones de conversión.
 * @param {boolean} options.enhanced - Si es true, convierte al formato mejorado.
 * @param {string} options.author - Autor para los metadatos en formato mejorado.
 * @returns {Object|null} - Objeto ATDF o null si hay error.
 */
function convertMcpFile(inputFile, outputFile = null, options = {}) {
  try {
    // Verificar existencia del archivo
    if (!fs.existsSync(inputFile)) {
      console.error(`Archivo no encontrado: ${inputFile}`);
      return null;
    }

    // Leer archivo JSON
    const mcpData = JSON.parse(fs.readFileSync(inputFile, 'utf8'));

    // Convertir a ATDF
    const atdfResult = mcpToAtdf(mcpData, options);

    // Guardar si se especificó archivo de salida
    if (outputFile) {
      // Crear directorio si no existe
      const outputDir = path.dirname(outputFile);
      if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
      }

      // Guardar el archivo
      fs.writeFileSync(outputFile, JSON.stringify(atdfResult, null, 2));
      console.log(`Herramienta guardada en ${outputFile}`);
    }

    return atdfResult;
  } catch (error) {
    console.error(`Error en la conversión: ${error.message}`);
    return null;
  }
}

/**
 * Procesa un lote de herramientas MCP.
 *
 * @param {Array|Object} mcpTools - Array de herramientas o objeto con propiedad 'tools'.
 * @param {string} outputDir - Directorio donde guardar las herramientas convertidas.
 * @param {Object} options - Opciones de conversión.
 * @returns {Array} - Array de herramientas ATDF convertidas.
 */
function batchConvert(mcpTools, outputDir, options = {}) {
  const results = [];

  // Normalizar entrada: puede ser un array o un objeto con propiedad 'tools'
  const toolsArray = Array.isArray(mcpTools) ? mcpTools : mcpTools.tools || [];

  // Crear directorio si no existe
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // Procesar cada herramienta
  for (let i = 0; i < toolsArray.length; i++) {
    const tool = toolsArray[i];
    try {
      // Convertir a ATDF
      const atdfTool = mcpToAtdf(tool, options);

      // Guardar en archivo
      const outputFile = path.join(outputDir, `${tool.name}.json`);
      fs.writeFileSync(outputFile, JSON.stringify(atdfTool, null, 2));

      console.log(
        `✅ Herramienta ${i + 1}/${toolsArray.length} convertida: ${outputFile}`
      );
      results.push(atdfTool);
    } catch (error) {
      console.error(
        `❌ Error al convertir herramienta ${tool.name}: ${error.message}`
      );
    }
  }

  return results;
}

// Exportar funciones
module.exports = {
  mcpToAtdf,
  convertMcpFile,
  batchConvert,
};
